ietfdata/tools/affiliations.py

```diff
@@ -0,0 +1,152 @@
+    def __str__(self):
+        return_str = '{"names":['
+        for name in self.names:
+            return_str+=f'"{name}",'
+        return_str =  return_str.rstrip(',')
+        return_str += "],"
+        return_str += f'"start_date":"{self.start_date}","end_date":"{self.end_date}"}}'
+        return return_str
+
+# Affiliation mapping class
+class AffiliationMap:
+    identifiers : list[str]
+    affiliations: list[AffiliationEntry]
+    
+    def __init__(self,identifiers:list[str],affiliations:list[AffiliationEntry]):
+        self.identifiers = copy.deepcopy(identifiers)
+        self.affiliations = copy.deepcopy(affiliations)
+    
+    def add_identifier(self, identifier:str):
+        self.identifiers.append(identifier)
+    
+    def add_affiliation(self, aff_entry:AffiliationEntry):
+        # TODO: Go through the timeline, insert the entry 
+        new_date = aff_entry.start_date
+        for affil in self.affiliations:
+            i = self.affiliations.index(affil)
+            if new_date <= affil.start_date:
+                self.affiliations.insert(i,aff_entry)
+                return #inserted entry
+        self.affiliations.append(aff_entry)
+        return
+            
+    def consolidate(self):
+        # TODO: Go through the timeline, consolidate the history
+        # This should only be run if and only if everything has been scraped
+        tmp_head_affil = None # temporary first affil in the batch
+        consolidated_affil = list()
+        for affil in self.affiliations:
+            if tmp_head_affil is None:
+                tmp_head_affil = copy.deepcopy(affil)
+                continue
+            # set_head_affil_names = set(tmp_head_affil.names)
+            # diff_list = [item for item in affil.names not in set_head_affil_names]
+            if not tmp_head_affil.match_names(affil.names):
+                    tmp_head_affil.end_date = (datetime.strptime(affil.start_date,'%Y-%m-%d').date() - timedelta(days=1))
+                    consolidated_affil.append(tmp_head_affil)
+                    tmp_head_affil = copy.deepcopy(affil)
+            if(tmp_head_affil not in consolidated_affil):
+                consolidated_affil.append(tmp_head_affil)
+        self.affiliations = copy.deepcopy(consolidated_affil)
+    
+    def __str__(self):
+        returnstr = '{"identifiers":['
+        for ident in self.identifiers:
+            returnstr += f'"{ident}",'
+        returnstr = returnstr[:-1] # strip last comma
+        returnstr += "],"
+        returnstr += '"affiliations":['
+        for affil in self.affiliations:
+            returnstr+=str(affil)
+            returnstr+=","
+        returnstr = returnstr[:-1] # strip last comma
+        returnstr += "]}"
+        return returnstr
+
+def remove_suffix(input_string:str, suffix:str):
+    if suffix and input_string.endswith(suffix):
+        return input_string[:-len(suffix)]
+    return input_string
+
+
+def cleanup_affiliation(affiliation_str:str):
```

Comment:
This does several different types of cleanup: it removes suffixes, such as “Ltd”, “Inc”, it normalises different spellings of “University”, and it runs through the large normalisation mapping list. Suggest to implement each of these as a separate function, then rewrite this function to call them in turn (this would make it easier to add new cleanup passes in future):
    ```
    def cleanup_affiliaton(affiliation:str) -> str:
      affiliation = cleanup_affiliation_suffix(affiliation)
      affiliation = cleanup_affiliation_academic(affiliation)
      affiliation = cleanup_affiliation_mist_affiliation)
      return affiliation
    ```